### ISSUE TITLE
Having some trouble training my Flappy Bird agent, looking for suggestions (and help understanding Tensorflow graphs)

### ISSUE THREAD

**Issue by ShaneCaler** [Unity-Technologies/ml-agents#1882]
Hey everyone, so I've been trying to train an agent to play Flappy Bird, and have been somewhat successful (it's gotten a high score of 34), but not as successful as I'd like. I'm wondering if you guys had any ideas on how I could improve my current system, and if you had any suggestions for hyperparameters as I'm a bit lost when it comes to them. I'll attach my code as paste bin links so it doesn't clutter this page, but I'll briefly describe my methodology below:
Collected Observations:
- 9 raycasts that return distance to the object hit ( a few pointing forward and then two that point up and down, that only interact with columns)
- agent's normalized Y position 
- agent's normalized Y velocity (for whatever reason I seem to be getting better results with these two values unnormalized, but from my understanding they should be normalized before being passed as an observation, correct?)
- X and Y distance from player to nearest goal object (should I normalize these two values before/after finding the distance?)

Discrete Actions:
- One action for adding jump force to player's rigidbody

Reward System:
- Small positive reward (.05f) each step for staying alive
- Positive reward (.5f) for passing through goal objects
- Negative reward (-1f) for colliding into floor, ceiling or a column (Using SetReward() here, is this the correct time to use that over AddReward()? Colliding also calls Done())

Training Hyperparameters:
- beta: 1.0e-2
- hidden_units: 256
- time_horizon: 32
- learning_rate: 5.0e-3
- batch_size: 512
- buffer_size: 20480

Agent Parameters:
- Max Steps: 0
- Decision Interval: 3 (a little confused on this parameter, not totally sure which situations would require a higher value and which would require a lower on. I figured since my agent needs to know pretty quickly if it's between two columns, so it should be lower - right?)

Academy Parameters:
Max Steps: 0
Time Scale: 100
While the bird does seem to learn, as you can see from the Tensorflow graphs below, rewards and training are all over the place. It will often crash on the first couple columns, but sometimes it will get lucky and hit a bunch in a row (i.e. the high score of 34).

Brain Parameters:
- 3 stacked vectors (experimenting with this, I assume that knowing the previous position of the columns would help the agent, right? So that it knows not to jump if it was between two columns the previous frame?)
- 1 discrete action, branch size of 2

Do you guys have any ideas on what I could change up to see more stable training? What should I look for in my Tensorflow graphs to gauge what exactly is going wrong? I would think that by 4 million steps, the agent would be able to get further than past 34 columns. The listed hyperparameters are just a product of guess-and-checking, so I'm not sure what works best yet.

Tensorflow Graphs after about 4,000,000 steps:
![ml1](https://user-images.githubusercontent.com/38741642/55377634-b21eca00-54ca-11e9-87a6-eae90689efd1.png)
![ml2](https://user-images.githubusercontent.com/38741642/55377638-b3e88d80-54ca-11e9-9f25-bb18417b1069.png)
![m3](https://user-images.githubusercontent.com/38741642/55377639-b519ba80-54ca-11e9-9925-31d1795201e9.png)

Pastebin links of code (I need to refactor and clean a lot of things, namely how my rayperception class is set up so I don't have to call Perceive a million times in CollectionObservations(), but that's for another day as long as it gets the job done):
- [BirdAgent.cs](https://pastebin.com/DCz5nsHE)
- [BirdAcademy.cs](https://pastebin.com/f6s4phY1)
- [BirdRayPerception.cs](https://pastebin.com/JmNt69BS)

Bonus gifs of the agent in training:

- [500,000 steps into training](https://imgur.com/a/liMf9WD)
- [1,000,000 steps into training](https://imgur.com/a/YR5dmFh)
- [2,000,000 steps into training](https://imgur.com/a/pdhPZ33)

**Comment by ervteng** [comment]
It looks like your agent is actually learning quite well. Do you know how far the bird moves per FixedUpdate? Decision Interval controls how frequently the agent takes an action. If this distance is quite far, then the agent may not be able to react in time to a goal that it sees - and I see that your bird is moving quite quickly. 

With a randomly generated level there will always be a bit of randomness in the reward. Does the level change at all as it progresses?

**Comment by mattinjersey** [comment]
what score would a human typically achieve on this game

**Comment by ShaneCaler** [comment]
> what score would a human typically achieve on this game

I have the spawn speed, scroll speed and distance between columns set to be super easy - I was just able to hit 35 without hardly trying, it just took forever because the scroll speed is so slow. Could have kept going but I lost patience lol, but the point is - it's much easier than the traditional flappy birds game is, though I would like to teach the agent to play harder versions later on.




> 
> 
> It looks like your agent is actually learning quite well. Do you know how far the bird moves per FixedUpdate? Decision Interval controls how frequently the agent takes an action. If this distance is quite far, then the agent may not be able to react in time to a goal that it sees - and I see that your bird is moving quite quickly.
> 
> With a randomly generated level there will always be a bit of randomness in the reward. Does the level change at all as it progresses?

In regard to the first question - sort of, currently the bird is fixed on it's x axis and just moves up and down, while the world scrolls past at a fixed rate (currently set to be very slow, as I mentioned above). I'll experiment with different values though and see what I come up with.

As for the second question, no not yet - the only changes that are happening is the spawn position for the columns, and it's basically just a small difference in Y values (for now, at least). The distance of the gap between the two, along with the spawn rate and world scroll rate are all fixed and pretty dang slow currently, which is why I would assume that the agent would be doing better by 4 million steps. 


Also, thanks for the replies! I know this post was pretty long-winded so I appreciate any help.

**Comment by ervteng** [comment]
Hi @ShaneCaler, hyperparameters are a bit of black magic sometimes, so definitely play around with them and see what happens. If your training appears to be unstable, I'd try lowering the learning rate.

**Comment by ervteng** [comment]
Closing due to inactivity. Feel free to open a new issue if you're still having problems!
